flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperHaServices.java

```diff
@@ -225,6 +227,12 @@ public void closeAndCleanupAllData() throws Exception {
 			exception = t;
 		}
 
+		try {
+			cleanupZooKeeperPaths();
```

Comment:
ZooKeeper has a method `ZKUtil#deleteRecursive` and Curator has a method `ZKPaths#deleteChildren`. Either or them can do the same thing as `#cleanupZooKeeperPaths`. We need not to re-implement it.

```java
  ZKPaths.deleteChildren(client.getZookeeperClient().getZooKeeper(), "/", true);
```

Reply:
I think `cleanupZooKeeperPaths` and `ZKPaths#deleteChildren` is not equivalent. On top of deleting the cluster id znode with all its children, we try to delete all parent nodes which might have been created when starting ZooKeeper. This should fail, if there is a parent with a remaining child because this indicates that this znode is still in use. Moreover, in `deleteOwnedZNode` we call `client.delete().deletingChildrenIfNeeded().forPath("/")` which is the same as `ZKPaths#deleteChildren`. The logic of deleting the parent nodes could be removed once we upgrade to ZooKeeper >= 3.5.1 which adds support for container nodes.

Reply:
You're right that without container mode we need `tryDeleteEmptyParentZNodes` to delete `/flink/clusterid`. 

However, if we wouldn't upgrade to ZooKeeper 3.5.x in the near future, it can be better that we parse path using `ZKPaths#getPathAndNode` instead of re-implement it. Besides, `CuratorFramework#getNamespace` always starts without a slash, we can directly add one or use `ZKPaths#makePath(client.getNamespace(), "")` to make it transparent.

I'm more curious that how we ensure that on a standby cluster shutdown, it doesn't accidentally delete all znodes under the running cluster id. Is it an impossible case?

Reply:
True, `ZKPaths#getPathAndNode` and `ZkPaths#makePath` look like good helper functions. I'll update the PR accordingly.

At the moment, I think the problem with stand-by `JobMasters` can occur. The `RunningJobsRegistry` was intended to solve this problem. However, at the moment, we clear everything once the `ExecutionGraph` reaches a globally terminal state. Instead what one could do is the following: 
- Only delete the `zNodes` not belonging to the running jobs registry if you are the leader transitioning the `ExecutionGraph` into a terminal state
- If you are the last `JobMaster` candidate which gains leadership, delete also the running jobs registry zNode. For that to work one needs to check the number of leader contenders.

I think this is actually a follow up issue because it is a bit bigger refactoring.

Reply:
For `RunningJobsRegistry` part, let's move the discussion to [FLINK-11813](https://issues.apache.org/jira/browse/FLINK-11813).

Besides, `#closeAndCleanupAllData` can be called if user sends a shutdown signal via REST. I'd like to know what if in this case can a standby cluster clean up ha data? Since we already close and clean up blob service, it might be not a problem.